File: examples/deadline/EC2-Image-Builder/ts/lib/template.ts
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from 'fs';

const DEFAULT_ENCODING = 'utf8';
const VALID_NAMES = /^[a-z][a-z0-9_]+$/i;

/**
 * Properties for invoking the template function
 */
export interface TemplateProps {
  /**
   * Path to the template file
   */
  readonly templatePath: string;

  /**
   * Mapping of token names to their substituted values.
   * Valid tokens are of the form `/^[a-z][a-z0-9_]+$/i`. Such as:
   * ```
   * ${NAME}
   * ```
   */
  readonly tokens: { [name: string]: string };

  /**
   * An optional encoding for the template file.
   *
   * @default "utf8"
   */
  readonly encoding?: BufferEncoding;
}

/**
 * Simple templating function. Loads a template from a path and substitutes all
 * occurrences of the tokens with their values.
 *
 * @param props The properties required to create the template
 * @returns The substituted template contents as a string
 */
export function templateComponent(props: TemplateProps): string {
  if (!props.templatePath.endsWith('.component.template')) {
    throw new Error(`Path does not end with ".component.template" ("${props.templatePath}")`);
  }

  const { templatePath, tokens } = props;
  const encoding = props.encoding ?? DEFAULT_ENCODING;

  Object.keys(tokens).forEach(name => {
    if (!VALID_NAMES.test(name)) {
      throw new Error(`Invalid token name "${name}"`);
    }
  });

  // Load the template from disk
  let result = fs.readFileSync(templatePath, { encoding });

  // Replace the tokens
  for (const [key, value] of Object.entries(tokens)) {
    result = result.replace(new RegExp(`\\\${${key}}`, 'g'), value);
  }

  return result;
}

File: examples/deadline/EC2-Image-Builder/python/package/lib/template.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from dataclasses import dataclass
import re

VALID_NAMES = re.compile(r"^[a-z][a-z0-9_]+$")

# Properties for invoking the template function
@dataclass
class TemplateProps():
    """
    Properties for an Image Builder component template
    """
    # Path to the template file.
    template_path: str

    # Mapping of token names to their substituted values.
    # Valid tokens are of the form `/^[a-z][a-z0-9_]+$/i`. Such as:
    # ```
    # ${NAME}
    # ```
    tokens: dict

# Simple templating function. Loads a template from a path and substitutes all
# occurrences of the tokens with their values. Tokens are of the form
def template_component(props: TemplateProps) -> str:
    if not props.template_path.endswith('.component.template'):
        raise Exception(f"Path does not end with \".component.template\": {props.template_path}")

    # Validate the tokens
    for name in props.tokens.keys():
        if VALID_NAMES.search(name) is None:
            raise Exception(f"Invalid token name {name}")

    f = open(props.template_path, "r")
    result = f.read()
    f.close()

    # Replace the tokens
    for name in props.tokens.keys():
        result = result.replace(f"${{{name}}}", props.tokens[name])

    return result
